crates/test-programs/wasi-sockets-tests/src/bin/tcp_v4.rs

```diff
@@ -15,6 +16,56 @@ fn wait(sub: poll::Pollable) {
     }
 }
 
+fn write(output: streams::OutputStream, bytes: &[u8]) -> (usize, StreamStatus) {
+    let sub = streams::subscribe_to_write_ready(output);
+    let num_wanted = loop {
+        wait(sub);
+        match streams::check_write(output) {
+            Some(WriteReadiness::Ready(num_wanted)) => break num_wanted,
+            Some(WriteReadiness::Closed) => return (0, StreamStatus::Ended),
+            None => (),
+        }
+    };
+    poll::drop_pollable(sub);
+
+    let num_to_write = min(num_wanted, bytes.len() as u64) as usize;
+    let bytes = &bytes[..num_to_write];
+
+    let mut num_written = 0;
+
+    while num_to_write != 0 {
+        let slice_to_write = &bytes[num_written..];
+        let num_written_this_iteration = match streams::write(output, slice_to_write) {
+            Some(WriteReadiness::Ready(num_written)) => num_written as usize,
+            Some(WriteReadiness::Closed) => return (num_written, StreamStatus::Ended),
+            None => {
+                match streams::flush(output) {
```

Comment:
`streams::blocking-flush` takes care of this entire nested case, because I figured it would be common.